src/Model/GContact.php

```diff
@@ -851,55 +851,137 @@ public static function update($contact)
-						'addr' => $contact['addr'], 'network' => $contact['network'],
-						'bd' => $contact['birthday'], 'gender' => $contact['gender'],
-						'keywords' => $contact['keywords'], 'alias' => $contact['alias'],
-						'contact-type' => $contact['contact-type'], 'url' => $contact['url'],
-						'location' => $contact['location'], 'about' => $contact['about']];
-
-				// Don't update the birthday field if not set or invalid
-				if (empty($contact['birthday']) || ($contact['birthday'] <= DBA::NULL_DATE)) {
-					unset($fields['bd']);
-				}
+		return $gcontact_id;
+	}
+
+	/**
+	 * @brief Updates the gcontact entry from a given public contact id
+	 *
+	 * @param integer $cid contact id
+	 * @return void
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 * @throws \ImagickException
+	 */
+	public static function updateFromPublicContactID($cid)
+	{
+		self::updateFromPublicContact(['id' => $cid]);
+	}
 
+	/**
+	 * @brief Updates the gcontact entry from a given public contact url
+	 *
+	 * @param string $url contact url
+	 * @return integer gcontact id
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 * @throws \ImagickException
+	 */
+	public static function updateFromPublicContactURL($url)
+	{
+		return self::updateFromPublicContact(['nurl' => Strings::normaliseLink($url)]);
+	}
 
-				DBA::update('contact', $fields, ['id' => $public_contact["id"]], $old_contact);
+	/**
+	 * @brief Helper function for updateFromPublicContactID and updateFromPublicContactURL
+	 *
+	 * @param array $condition contact condition
+	 * @return integer gcontact id
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 * @throws \ImagickException
+	 */
+	private static function updateFromPublicContact($condition)
+	{
+		$fields = ['name', 'nick', 'url', 'nurl', 'location', 'about', 'keywords', 'gender',
+			'bd', 'contact-type', 'network', 'addr', 'notify', 'alias', 'archive', 'term-date',
+			'created', 'updated', 'avatar', 'success_update', 'failure_update', 'forum', 'prv',
+			'baseurl', 'sensitive', 'unsearchable'];
+
+		$contact = DBA::selectFirst('contact', $fields, array_merge($condition, ['uid' => 0, 'network' => Protocol::FEDERATED]));
+		if (!DBA::isResult($contact)) {
+			return 0;
+		}
+
+		$fields = ['name', 'nick', 'url', 'nurl', 'location', 'about', 'keywords', 'gender', 'generation',
+			'birthday', 'contact-type', 'network', 'addr', 'notify', 'alias', 'archived', 'archive_date',
+			'created', 'updated', 'photo', 'last_contact', 'last_failure', 'community', 'connect',
+			'server_url', 'nsfw', 'hide', 'id'];
+
+		$old_gcontact = DBA::selectFirst('gcontact', $fields, ['nurl' => $contact['nurl']]);
+		$do_insert = !DBA::isResult($old_gcontact);
+		if ($do_insert) {
+			$old_gcontact = [];
+		}
+
+		$gcontact = $contact;
+
+		// These fields are having different names but the same content
+		$gcontact['server_url'] = $gcontact['baseurl'];
```

Comment:
Can you please untangle the variable assignments and the `unset` calls in two separate blocks? This proves difficult to read for me.

Reply:
I made it this way, so that I don't forget something. When all assignments and unset are put together, then it is more likely that an unset is forgotten, I think.

Reply:
Another idea, having the unset(s) on the same line: 
```php
$gcontact['server_url'] = $gcontact['baseurl']; unset($gcontact['baseurl']);
```

Reply:
Another idea: only assigning the `$gcontact` fields you need from the `$contact` variable. No unsets.

Reply:
This would be a really long list. I find those lists mostly ugly. I have to see which is looking better.